File: tool/feedback/src/main/java/feedback/parser/LogEntryBuilder.java
package feedback.parser;

import org.joda.time.DateTime;

import java.io.Serializable;

final class LogEntryBuilder implements Serializable {
    public final DateTime datetime;
    public final String type;
    public final String thread;
    public final String file;
    public final int fileLine;

    private int logLine;
    private final StringBuilder msg;

    LogEntryBuilder(String datetimeText, String logTypeText, String locationText, String msg) {
        this.datetime = Parser.parseDatetime(datetimeText);
        this.type = Parser.parseLogType(logTypeText);
        final scala.Tuple3<String, String, Integer> location = Parser.parseLocation(locationText);
        this.thread = location._1();
        this.file = location._2();
        this.fileLine = location._3();
        this.logLine = -1;    // should be set later
        this.msg = new StringBuilder(msg);
    }

    void setLogLine(int logLine) {
        this.logLine = logLine;
    }

    int getLogLine() {
        return this.logLine;
    }

    void appendNewLine(String s) {
        this.msg.append('\n').append(s);
    }

    LogEntry build() {
        if (logLine == -1) {
            throw new RuntimeException("Bad log line");
        }
        return new LogEntry(datetime, type, thread, file, fileLine, msg.toString(), logLine);
    }

    String getMsg() {
        return msg.toString();
    }

    LogEntry buildWithoutLogLine() {
        return new LogEntry(datetime, type, thread, file, fileLine, msg.toString(), -1);
    }
}

File: tool/feedback/src/main/scala/feedback/parser/LogEntryBuilder.scala
package feedback.parser

import org.joda.time.DateTime

import scala.collection.mutable

private[parser] final class LogEntryBuilder(val datetime: DateTime,
                                            val logType: LogType,
                                            val thread: String,
                                            val file: String,
                                            val fileLine: Int,
                                            private var logLine: Int,
                                            private var msg: mutable.StringBuilder) extends java.io.Serializable {

  private def this(datetime: DateTime, logType: LogType, thread: String, file: String, fileLine: Int, msg: String) =
    this(datetime, logType, thread, file, fileLine, -1, new mutable.StringBuilder(msg))

  private def this(datetime: String, logType: String, thread: String, file: String, fileLine: Int, msg: String) =
    this(Parser.parseDatetime(datetime), Parser.parseLogType(logType), thread, file, fileLine, msg)

  def setLogLine(logLine: Int): Unit = this.logLine = logLine
  def getLogLine: Int = this.logLine

  def appendNewLine(text: String): Unit = this.msg.append('\n').append(text)

  def build: LogEntry =
    if (logLine == -1) throw new RuntimeException("Bad log line")
    else LogEntryBuilder.build(datetime, logType, thread, file, fileLine, msg.toString, logLine)

  def buildWithoutLogLine: LogEntry = LogEntryBuilder.build(datetime, logType, thread, file, fileLine, msg.toString, -1)

  def getMsg: String = this.msg.toString
  def resetMsg(msg: String): mutable.StringBuilder = {
    this.msg = new mutable.StringBuilder(msg)
    this.msg
  }
}

private[parser] object LogEntryBuilder {
  def create(datetimeText: String, logTypeText: String, locationText: String, msg: String): LogEntryBuilder =
    Parser.parseLocation(locationText) match {
      case (thread, file, fileLine) =>
        new LogEntryBuilder(datetimeText, logTypeText, thread, file, fileLine, msg)
    }

  private def build(datetime: DateTime, logType: LogType, thread: String, file: String, fileLine: Int, text: String, logLine: Int): LogEntry =
    Parser.parseLogException(text) match {
      case (msg, exceptions) => new LogEntry(datetime, logType, thread, file, fileLine, msg, logLine, exceptions)
    }
}
